### ISSUE TITLE
Crash on anonymous type parameters of implicit classes

### ISSUE THREAD

**Issue by xeno-by** [twitter/rsc#366]
```scala
object M {
  implicit class C[_](x: Int) {
    def m: Int = ???
  }
}
```

```
error: compiler crash: None.get
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:347)
	at scala.None$.get(Option.scala:345)
	at rsc.outline.Synthesizer$$anonfun$14.rsc$outline$Synthesizer$$anonfun$$synthesize$1(Synthesizer.scala:157)
	at rsc.outline.Synthesizer$$anonfun$14$$anonfun$15.apply(Synthesizer.scala:158)
	at rsc.outline.Synthesizer$$anonfun$14$$anonfun$15.apply(Synthesizer.scala:158)
	at scala.collection.immutable.List.map(List.scala:284)
	at rsc.outline.Synthesizer$$anonfun$14.rsc$outline$Synthesizer$$anonfun$$synthesize$1(Synthesizer.scala:158)
	at rsc.outline.Synthesizer$$anonfun$14.apply(Synthesizer.scala:164)
	at rsc.outline.Synthesizer$$anonfun$14.apply(Synthesizer.scala:154)
	at scala.collection.immutable.List.map(List.scala:284)
	at rsc.outline.Synthesizer.implicitClassConversion(Synthesizer.scala:154)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:515)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:503)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.outline.Scheduler.stats(Scheduler.scala:503)
	at rsc.outline.Scheduler.defnTemplate(Scheduler.scala:341)
	at rsc.outline.Scheduler.apply(Scheduler.scala:35)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:512)
	at rsc.outline.Scheduler$$anonfun$stats$2.apply(Scheduler.scala:503)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.outline.Scheduler.stats(Scheduler.scala:503)
	at rsc.outline.Scheduler.source(Scheduler.scala:470)
	at rsc.outline.Scheduler.apply(Scheduler.scala:38)
	at rsc.Compiler$$anonfun$rsc$Compiler$$schedule$1.apply(Compiler.scala:108)
	at rsc.Compiler$$anonfun$rsc$Compiler$$schedule$1.apply(Compiler.scala:106)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at rsc.Compiler.rsc$Compiler$$schedule(Compiler.scala:106)
	at rsc.Compiler$$anonfun$phases$3.apply$mcV$sp(Compiler.scala:70)
	at rsc.Compiler$$anonfun$run$2.apply(Compiler.scala:36)
	at rsc.Compiler$$anonfun$run$2.apply(Compiler.scala:33)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at rsc.Compiler.run(Compiler.scala:33)
	at rsc.cli.Main$.process(Main.scala:35)
	at rsc.cli.Main$.main(Main.scala:14)
	at rsc.cli.Main.main(Main.scala)
```

**Comment by xeno-by** [comment]
This code has pretty weird behavior, so I'm not gonna reproduce it right away, especially since our case study project doesn't actually rely on it:

```
$ s -Xprint:typer
[[syntax trees at end of                     typer]] // C.scala
package <empty> {
  object M extends scala.AnyRef {
    def <init>(): M.type = {
      M.super.<init>();
      ()
    };
    implicit class C[_] extends scala.AnyRef {
      <paramaccessor> private[this] val x: Int = _;
      def <init>(x: Int): M.C[_] = {
        C.super.<init>();
        ()
      };
      def m: Int = scala.this.Predef.???
    };
    implicit <synthetic> def C[_](x: Int): M.C[Any] = new M.C[Any](x)
  }
}
```

**Comment by sundresh** [comment]
I looked into this a bit to get a better idea of how `rsc` works.

For one, if we remove the `implicit` keyword, the compiler does not crash. However, dumping the generated outline class file with `scalap`, the `_` is converted to `anon$1`. This causes the following code to crash `rsc`, while it compiles with `scalac`:
```
object M {
  class C[_, anon$1](x: Int) {
    def m: Int = ???
  }
}
```
(Using `anon$2` instead of `anon$1` above does not cause `rsc` to crash.)

Another interesting thing is the following code compiles with `rsc` (with the two `_`'s being replaced by `anon$1` and `anon$2`), while `scalac` 2.11.11 returns an error:
```
object M {
  class C[_, _](x: Int) {
    def m: Int = ???
  }
}
```
```
$ scalac M.scala
M.scala:2: error: _ is already defined as type _
    class C[_, _](x: Int) {
               ^
one error found
```

It seems that a direct `_` type parameter in a class or method definition is handled almost exactly like a type variable, except that no binding is introduced for that type variable within the scope of the class. Perhaps we need to adjust the parser to differentiate between this case and higher kinded type parameters (like `class C[Map[_, _]] { ... }`).

----

Now, the reason that the `implicit class` is handled differently from a plain `class` is there's a `match`-`case` in `Scheduler.stats()` that specifically looks for an `implicit class`, and when found, calls `synthesizer.implicitClassConversion(currEnv, tree)`. The crash this ticket describes is at some point within this call.
```
        tree match {
          case tree: DefnClass if tree.hasImplicit =>
            synthesizer.implicitClassConversion(currEnv, tree)
          case _: DefnCtor | _: PrimaryCtor =>
            ()
          case tree: DefnDef if tree.hasDefaultParams =>
            synthesizer.defaultGetters(treeEnv, tree)
          case _ =>
            ()
        }
```

So the problem is at least partly in `rsc.outline.Synthesizer`. I noticed a `FIXME` comment at the top of that class that points to #98, so overall it seems the `Synthesizer` class needs some improvement.

**Comment by sundresh** [comment]
I learned a bit more about what's going on here. The reason implicit classes are handled differently is we need to synthesize an implicit factory method for them (that's what `implicitClassConversion()` does). In the example above, the factory method generated by `scalac` would be:

```
implicit def C[_](x: Int): C[Any] = new C[_](x)
```

The reason the program crashes is the type cast in the following snippet from `implicitClassConversion()` fails to cast an `AnonId` (which represents `_`), since it is not a subclass of `TptId` (I guess because `_` can also be used in non-type contexts?)

```
      if (tparams.isEmpty) Some(core)
      else Some(TptParameterize(core, tparams.map(_.id.asInstanceOf[TptId])))
```

The reason the code above exists is to convert the type parameters of an implicit class to the return type parameters of the implicit factory method. The type parameters of the implicit class can include things like upper and lower bounds and `_`, while the return type parameters can't (they can only include concrete types and type variables).

What the code _should_ do is use `match`-`case` instead of `.asInstanceOf`, and convert `AnonId` to a `TptPath` representing `_root_.scala.Any`.

I'm working on a fix implementing what I described in this message + comments + a test. We should probably open a separate ticket for the other discrepancies I noticed in my message above.
